src/otx/v2/adapters/torch/dataset.py

```diff
@@ -0,0 +1,210 @@
+"""OTX adapters.torch.mmengine.mmpretrain.Dataset API."""
+
+# Copyright (C) 2023 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+from __future__ import annotations
+
+from pathlib import Path
+from typing import Iterable
+
+import yaml
+from torch.utils.data import DataLoader as TorchDataLoader
+from torch.utils.data import Dataset as TorchDataset
+from torch.utils.data import Sampler
+
+from otx.v2.api.core.dataset import BaseDataset
+from otx.v2.api.utils import set_tuple_constructor
+from otx.v2.api.utils.decorators import add_subset_dataloader
+
+SUBSET_LIST = ["train", "val", "test"]
+
+
+@add_subset_dataloader(SUBSET_LIST)
+class TorchBaseDataset(BaseDataset):
```

Comment:
It seems strange to me that `Dataset` builds `Dataset` (itself?) and `DataLoader`. Pytorch Lightning's `DataModule` design looks much more natural https://lightning.ai/docs/pytorch/stable/data/datamodule.html.

Since we decided to go with Pytorch Lightning based framework, why not introduce the `DataModule` concept here?

Reply:
Yes, if we change to a lightning based framework I agree. I'm still not sure about the decision.

Anyway, @samet-akcay As we feared, the name Dataset seems to cause a lot of confusion for developers.
Maybe we should rename it to something like DatasetBuilder. What do you think?

Reply:
DataModule seems fine

Reply:
Vote for `DataModule`.

Reply:
I agree that the behaviour of this class is very similar to a Lightning DataModule. The only problem I see with calling it this way is that we would then have a class that is called `adapters.torch.Lightning.DataModule` but which is not actually an instance of a Lightning DataModule. Not sure how problematic this would be

Reply:
We'll continue the discussion and refactoring of this after this PR is merged.

Reply:
we could discuss it during the next phase, but I agree with @djdameln. I feel that `DataModule` would be confusing as `DataModule` is known to be a lightning component. Having our `DataModule` in Lightning adapters might potentially confuse users. 

Based on this argument, `Dataset` might also be confusing as torch already has this :) 

Anyway, leaving these here for a future reference. When the time comes, we could discuss this again.